src/server/project.ts

```diff
@@ -1068,78 +1135,83 @@ namespace ts.server {
             }));
         }
 
-        watchConfigFile(callback: (project: ConfiguredProject) => void) {
-            this.projectFileWatcher = this.projectService.host.watchFile(this.getConfigFilePath(), _ => callback(this));
-        }
-
-        watchTypeRoots(callback: (project: ConfiguredProject, path: string) => void) {
-            const roots = this.getEffectiveTypeRoots();
-            const watchers: FileWatcher[] = [];
-            for (const root of roots) {
-                this.projectService.logger.info(`Add type root watcher for: ${root}`);
-                watchers.push(this.projectService.host.watchDirectory(root, path => callback(this, path), /*recursive*/ false));
-            }
-            this.typeRootsWatchers = watchers;
-        }
-
-        watchConfigDirectory(callback: (project: ConfiguredProject, path: string) => void) {
-            if (this.directoryWatcher) {
-                return;
-            }
-
-            const directoryToWatch = getDirectoryPath(this.getConfigFilePath());
-            this.projectService.logger.info(`Add recursive watcher for: ${directoryToWatch}`);
-            this.directoryWatcher = this.projectService.host.watchDirectory(directoryToWatch, path => callback(this, path), /*recursive*/ true);
-        }
-
-        watchWildcards(callback: (project: ConfiguredProject, path: string) => void) {
-            if (!this.wildcardDirectories) {
-                return;
-            }
-            const configDirectoryPath = getDirectoryPath(this.getConfigFilePath());
-
-            this.directoriesWatchedForWildcards = createMap<FileWatcher>();
-            this.wildcardDirectories.forEach((flag, directory) => {
-                if (comparePaths(configDirectoryPath, directory, ".", !this.projectService.host.useCaseSensitiveFileNames) !== Comparison.EqualTo) {
+        watchWildcards(wildcardDirectories: Map<WatchDirectoryFlags>) {
+            this.directoriesWatchedForWildcards = mutateExistingMap(

```

Comment:
It would help readability if `mutateExistingMap` took an object literal with options, instead of taking so many positional parameters. Same for `mutateExistingMapWithNewSet`.